### PR TITLE
Pv2 932 inconsistent provider behaviour   changes to the required payments service

### PR DESCRIPTION
**Pull request by abroodie**

### CODE REVIEW

File: src/SFA.DAS.Payments.RequiredPayments.Domain/Services/ApprenticeshipKeyService.cs [hunk 1]
```diff
@@ -53,5 +55,12 @@ private long ParseToLong(string source, string destinationName)
                 throw new ArgumentException($"Cannot parse the key part to a long. Destination: {destinationName}, source value: '{source}'");
             return result;
         }
+
+        private short ParseToShort(string source, string destinationName)
+        {
+            if (!short.TryParse(source, out var result))
+                throw new ArgumentException($"Cannot parse the key part to a long. Destination: {destinationName}, source value: '{source}'");
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.Domain/Services/ApprenticeshipKeyService.cs]: done

File: src/SFA.DAS.Payments.RequiredPayments.Application/Extensions/PeriodisedRequiredPaymentEventExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using FluentAssertions;
+using SFA.DAS.Payments.RequiredPayments.Messages.Events;
+using SFA.DAS.Payments.RequiredPayments.Model.Entities;
+
+namespace SFA.DAS.Payments.RequiredPayments.Application.Extensions
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Extensions/PeriodisedRequiredPaymentEventExtensions.cs]: done

File: src/SFA.DAS.Payments.RequiredPayments.Application/Processors/PeriodisedRequiredPaymentEventFactory.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                    {
+                        Type = (IncentivePaymentType) transactionType,
+                    };
+                case EarningType.Levy:
+                    return new CalculatedRequiredLevyAmount
+                    {
+                        OnProgrammeEarningType = (OnProgrammeEarningType) transactionType,
+                    };
+                default:
+                    throw new InvalidOperationException($"Unknown earning type found: {earningType:G}. Cannot create the PeriodisedRequiredPaymentEvent.");
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Processors/PeriodisedRequiredPaymentEventFactory.cs]: enum camel case

File: src/SFA.DAS.Payments.RequiredPayments.Application/Processors/RefundRemovedLearningAimProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                throw new InvalidOperationException("No payment history found in the cache.");
+            var historicPayments = cacheItem.Value.Select(mapper.Map<PaymentHistoryEntity, Payment>).ToList();
+            logger.LogDebug($"Got {historicPayments.Count} historic payments. Now generating refunds per transaction type.");
+
+            var requiredPaymentEvents = historicPayments.GroupBy(historicPayment => historicPayment.TransactionType)
+                .SelectMany(group =>
+                    refundRemovedLearningAimService.RefundLearningAim(historicPayments)
+                        .Select(requiredPayment =>
+                        {
+                            logger.LogVerbose($"Now mapping the required payment to a PeriodisedRequiredPaymentEvent.");
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Processors/RefundRemovedLearningAimProcessor.cs]: not really, it isn't that many lines of code

File: src/SFA.DAS.Payments.RequiredPayments.Domain.UnitTests/Services/RefundRemovedLearningAimServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        [Test]
+        public void Refunds_All_Delivery_Periods()
+        {
+            mocker.Mock<IRefundService>().Setup(x => x.GetRefund(It.IsAny<decimal>(), It.IsAny<List<Payment>>()))
+                .Returns(new List<RequiredPayment>());
+
+            history.AddRange(Enumerable.Range(1, 12).Select(period => new Payment { DeliveryPeriod = (byte)period, SfaContributionPercentage = .9M, Amount = period * 10, PriceEpisodeIdentifier = "pe-1" }));
+            var service = mocker.Create<RefundRemovedLearningAimService>();
+            var requiredPayments = service.RefundLearningAim(history);
+            //requiredPayments.Count.Should().Be(12);
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.Domain.UnitTests/Services/RefundRemovedLearningAimServiceTests.cs]: done

File: src/SFA.DAS.Payments.RequiredPayments.Domain.UnitTests/Services/RefundRemovedLearningAimServiceTests.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            Enumerable.Range(1, 12).ToList().ForEach(period => mocker.Mock<IRefundService>().Verify(x => x.GetRefund(It.Is<decimal>(amount => amount == 0M), It.Is<List<Payment>>(payments => payments.All(payment => payment.DeliveryPeriod == period))),Times.Once));
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.Domain.UnitTests/Services/RefundRemovedLearningAimServiceTests.cs]: made simpler

File: src/SFA.DAS.Payments.RequiredPayments.Domain/Services/RefundRemovedLearningAimService.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        public RefundRemovedLearningAimService(IRefundService refundService)
+        {
+            this.refundService = refundService ?? throw new ArgumentNullException(nameof(refundService));
+        }
+
+        public List<RequiredPayment> RefundLearningAim(List<Payment> historicPayments)
+        {
+            return historicPayments
+                .GroupBy(payment => payment.DeliveryPeriod)
+                .SelectMany(group => refundService.GetRefund(0, group.ToList()))
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.RequiredPayments.Domain/Services/RefundRemovedLearningAimService.cs]: What is this 0?

**Pull request merged**